core/sql/optimizer/NodeMap.cpp

```diff
@@ -1311,6 +1311,135 @@ NodeMap::getPopularNodeNumber(CollIndex beginPos, CollIndex endPos) const
   return popularNodeNumber;
 } // NodeMap::getNodeNum
 
+// Smooth the node map to reassign entries with identical node Id of higher
+// frequency than the rest to some other nodes. Assume there are m total entries in the map, and 
+// n nodes in the cluster. Assume further there are s entries in the map (out of m) that refer 
+// to very few nodes.  These s entrres are the subject of smoothing operation.  We do so by
+// 1. allocate an array nodeUsageMap[] and the ith entry in it contains all indexes k in the map
+//    that points at i (i.e., nodeMap.getNodeId[k] = i)
+// 2. Find out s by frequency counting
+// 3. Find out m - s
+// 4. Find out how many nodes in each of the s entries that should be moved
+// 
+// The nodels accepting the reasssignment will be from the set of the nodes contained in the map.
+//
+// Example 1.  Assume the original node map with 6 entries as follows:
+//
+//    NodeMapIndex   0   1   2   3  4  5 
+//    NodeMapEntry   0   1   1   3  1  5
+//
+// The nodes accepting the reasssignment will be { 0, 3, 5 }
+//
+// The subset of entries referring to a few nodes with high frequency: s = { 1, 2, 4 } 
+// Since f(0)=f(3)=f(5)=1, the average frequency of nodes not in s: (1+1+1)/3 = 1.
+// We will allow one (1) assigment in s to remain in node 1 since it is the average frequency, and 
+// re-assign the rest (marked X) to different nodes via round-robin starting the 1st node in the map. 
+// The node (1) is excluded from the re-assignment. 
+//
+//    NodeMapIndex   0   1   2   3  4  5 
+//    NodeMapEntry   0   1   1   3  1  5
+//                           X      X
+//                           ^      ^
+//                           |      |
+//                          to 0   to 3
```

Comment:
It is included for number of times up to "base frequencies". Beyond that,
other nodes are used.

On Fri, Sep 25, 2015 at 12:18 PM, Suresh Subbiah <notifications@github.com>
wrote:

> In core/sql/optimizer/NodeMap.cpp
> <https://github.com/apache/incubator-trafodion/pull/92#discussion_r40453224>
> :
>
> > +//    NodeMapEntry   0   1   1   3  1  5
> > +//
> > +// The nodes accepting the reasssignment will be { 0, 3, 5 }
> > +//
> > +// The subset of entries referring to a few nodes with high frequency: s = { 1, 2, 4 }
> > +// Since f(0)=f(3)=f(5)=1, the average frequency of nodes not in s: (1+1+1)/3 = 1.
> > +// We will allow one (1) assigment in s to remain in node 1 since it is the average frequency, and
> > +// re-assign the rest (marked X) to different nodes via round-robin starting the 1st node in the map.
> > +// The node (1) is excluded from the re-assignment.
> > +//
> > +//    NodeMapIndex   0   1   2   3  4  5
> > +//    NodeMapEntry   0   1   1   3  1  5
> > +//                           X      X
> > +//                           ^      ^
> > +//                           |      |
> > +//                          to 0   to 3
>
> Should node map entry = 2 be included?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/92/files#r40453224>.
>



-- 
Regards, --Qifan